### PR TITLE
Update member profile & Get/Update Diary Settings APIs

### PR DESCRIPTION
**Pull request by peteclark3**
https://github.com/Open-Diary/opendiary/issues/161
https://github.com/Open-Diary/opendiary/issues/162
https://github.com/Open-Diary/opendiary/issues/163

- This is now custom to our site, so I ditched the old array(...) etc syntax.

### CODE REVIEW

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 1]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
-				'args'                => array(
-					'context' => $this->get_context_param( array( 'default' => 'view' ) ),
-				),
-			),
-			'schema' => array( $this, 'get_public_item_schema' ),
-		) );
+				'callback'            => [ $this, 'get_item' ],
+				'permission_callback' => [ $this, 'get_item_permissions_check' ],
+				'args'                => [
+					'context' => $this->get_context_param( [ 'default' => 'view' ] ),
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:74]: fix spacing here.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 2]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+	 *
+	 * @since 0.0.1
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or error object on failure.
+	 */
+	public function update_item( $request ) {
+		$id = (int) $request['id'];
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php]: `$id = intval( $request['id'] );`

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 3]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+
+		$user = get_user_by( "id", $id );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php]: single quote to stay consistent.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 4]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+			xprofile_set_field_data( self::FIELD_ABOUT_ME_XPROFILE, $id, $prepared_args[ self::FIELD_ABOUT_ME_API ] );
+		}
+		if ( array_key_exists( self::FIELD_GENDER_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_GENDER_XPROFILE, $id, $prepared_args[ self::FIELD_GENDER_API ] );
+		}
+		if ( array_key_exists( self::FIELD_DOB_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DOB_XPROFILE, $id, $prepared_args[ self::FIELD_DOB_API ] );
+		}
+		if ( array_key_exists( self::FIELD_COUNTRY_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_COUNTRY_XPROFILE, $id, $prepared_args[ self::FIELD_COUNTRY_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:128]: Check `OD\Helpers` for method to check against valid countries.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 5]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+			xprofile_set_field_data( self::FIELD_NAME_XPROFILE, $id, $prepared_args[ self::FIELD_NAME_API ] );
+		}
+		if ( array_key_exists( self::FIELD_ABOUT_ME_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_ABOUT_ME_XPROFILE, $id, $prepared_args[ self::FIELD_ABOUT_ME_API ] );
+		}
+		if ( array_key_exists( self::FIELD_GENDER_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_GENDER_XPROFILE, $id, $prepared_args[ self::FIELD_GENDER_API ] );
+		}
+		if ( array_key_exists( self::FIELD_DOB_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DOB_XPROFILE, $id, $prepared_args[ self::FIELD_DOB_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:125]: Need to validate data is correct.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 6]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+
+		if ( array_key_exists( self::FIELD_NAME_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_NAME_XPROFILE, $id, $prepared_args[ self::FIELD_NAME_API ] );
+		}
+		if ( array_key_exists( self::FIELD_ABOUT_ME_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_ABOUT_ME_XPROFILE, $id, $prepared_args[ self::FIELD_ABOUT_ME_API ] );
+		}
+		if ( array_key_exists( self::FIELD_GENDER_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_GENDER_XPROFILE, $id, $prepared_args[ self::FIELD_GENDER_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:122]: Check `OD\Helpers` for method to check against valid genders.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 7]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+		if ( array_key_exists( self::FIELD_DOB_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DOB_XPROFILE, $id, $prepared_args[ self::FIELD_DOB_API ] );
+		}
+		if ( array_key_exists( self::FIELD_COUNTRY_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_COUNTRY_XPROFILE, $id, $prepared_args[ self::FIELD_COUNTRY_API ] );
+		}
+		if ( array_key_exists( self::FIELD_STATEPROV_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_STATEPROV_XPROFILE, $id, $prepared_args[ self::FIELD_STATEPROV_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:131]: Check `OD\Helpers` for method to check against valid states.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 8]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+		if ( array_key_exists( self::FIELD_DIARY_TITLE_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DIARY_TITLE_XPROFILE, $id, $prepared_args[ self::FIELD_DIARY_TITLE_API ] );
+		}
+		if ( array_key_exists( self::FIELD_DIARY_DESC_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DIARY_DESC_XPROFILE, $id, $prepared_args[ self::FIELD_DIARY_DESC_API ] );
+		}
+		if ( array_key_exists( self::FIELD_DIARY_PRIVACY_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DIARY_PRIVACY_XPROFILE, $id, $prepared_args[ self::FIELD_DIARY_PRIVACY_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:140]: Need to validate privacy (might be a helper, if not add it in).

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 9]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+		if ( array_key_exists( self::FIELD_DIARY_ALLOW_COMMENTS_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DIARY_ALLOW_COMMENTS_XPROFILE, $id, $prepared_args[ self::FIELD_DIARY_ALLOW_COMMENTS_XPROFILE ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:143]: Validate allow comments (might be a helper, if not add it in).

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 10]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+		if ( array_key_exists( self::FIELD_DIARY_ALLOW_PRIV_DIARY_COMMENT_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DIARY_ALLOW_PRIV_DIARY_COMMENT_XPROFILE, $id, $prepared_args[ self::FIELD_DIARY_ALLOW_PRIV_DIARY_COMMENT_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:146]: I believe this is boolean, but needs to be validated.

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 11]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+		if ( array_key_exists( self::FIELD_DIARY_ALLOW_PRIV_COMMENT_ON_DIARY_API, $prepared_args ) ) {
+			xprofile_set_field_data( self::FIELD_DIARY_ALLOW_PRIV_COMMENT_ON_DIARY_XPROFILE, $id, $prepared_args[ self::FIELD_DIARY_ALLOW_PRIV_COMMENT_ON_DIARY_API ] );
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php:149]: Validate boolean (I think it's boolean...)

File: includes/bp-members/classes/class-bp-members-endpoints.php [hunk 12]
```diff
@@ -24,27 +55,175 @@ public function __construct() {
...
+		}
+
+		$request->set_param( 'context', 'edit' );
+
+		$updated_user_profile = $this->_get_user_profile_datum( [
+			'include' => [ (int) $request['id'] ],
```
Comment (mauteri) [inline includes/bp-members/classes/class-bp-members-endpoints.php]: `intval( $request['id'] )`

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 1 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:74]: This was what PhpStorm's wordpress formatting did; I'll make it better on my own.

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 4 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:128]: SWEET

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 5 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:125]: Can you elaborate on "correct?"  - there is sanitization/format checking happening in the prepare for db function

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 10 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:146]: Values are "Yes" and "No" according to xprofile

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 11 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:149]: Values are "Yes" and "No" according to xprofile

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 1 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:74]: Seems ok in my ide...

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 2 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php]: done

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 3 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php]: done

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 4 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:128]: done

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 5 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:125]: i think you meant 'date' , not data...  validiating format now

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 6 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:122]: done

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 7 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:131]: done

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 8 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:140]: done and added

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 9 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:143]: done and added

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 10 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:146]: done and added

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 11 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php:149]: done and added

File: includes/bp-members/classes/class-bp-members-endpoints.php (hunk 12 above)
Comment (peteclark3) [reply includes/bp-members/classes/class-bp-members-endpoints.php]: done

**Pull request merged**